lib/ClangImporter/ClangImporter.cpp

```diff
@@ -2256,6 +2256,120 @@ ClangModuleUnit *ClangImporter::Implementation::getClangModuleForDecl(
+    const SwiftLookupTable::SingleEntry &target) {
+  assert(target);
+  // In some situations, the ClangImporter is invoked in a reentrant manner,
+  // resulting in a diagnosticTarget already being set. In such a case, we
+  // keep the outer diagnostic target.
+  if (!diagnosticTarget) {
+    diagnosticTarget = target;
+    return true;
+  }
+  return false;
+}
+
+SwiftLookupTable::SingleEntry
+ClangImporter::Implementation::getDiagnosticTarget() {
+  return diagnosticTarget;
+}
+
+void ClangImporter::Implementation::resetDiagnosticTarget(bool shouldReset) {
+  if (shouldReset) {
+    assert(diagnosticTarget);
+    diagnosticTarget = nullptr;
+  }
+}
+
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const SwiftLookupTable::SingleEntry &targetCandidate) const {
+  return diagnosticTarget == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::NamedDecl *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::NamedDecl *>() == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::ModuleMacro *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::ModuleMacro *>() == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::MacroInfo *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::MacroInfo *>() == targetCandidate;
+}
+
+bool ClangImporter::Implementation::setEagerImportActive() {
+  bool shouldReset = !eagerImportActive;
+  eagerImportActive = true;
+  return shouldReset;
+}
+
+void ClangImporter::Implementation::resetEagerImportActive(bool shouldReset) {
+  if (shouldReset) {
+    eagerImportActive = false;
+  }
+}
+
+bool ClangImporter::Implementation::getEagerImportActive() {
+  return eagerImportActive;
+}
+
+void ClangImporter::Implementation::reportNoteConsumingDiagnosticError(
+    const clang::NamedDecl *target, Diagnostic &&error,
+    const clang::SourceLocation &loc) {
+  if (SwiftContext.LangOpts.EnableExperimentalClangImporterDiagnostics &&
+      isDiagnosticTarget(target)) {
+    diagnose(convertSourceLocation(loc), error);
+    for (auto noteIt = pendingErrorNotes.rbegin();
+         noteIt != pendingErrorNotes.rend(); ++noteIt) {
+      if (!noteIt->diag.getLoc()) {
+        noteIt->diag.setLoc(convertSourceLocation(loc));
+      }
+      if (isDiagnosticTarget(noteIt->target)) {
+        diagnose(noteIt->diag.getLoc(), noteIt->diag);
+      }
+    }
+  }
+  pendingErrorNotes.clear();
+}
+
+void ClangImporter::Implementation::addPendingErrorNote(Diagnostic &&note) {
+  addPendingErrorNote(std::move(note), clang::SourceLocation());
+}
+
+void ClangImporter::Implementation::addPendingErrorNote(
+    Diagnostic &&note, const clang::SourceLocation &loc) {
+  if (SwiftContext.LangOpts.EnableExperimentalClangImporterDiagnostics &&
+      getDiagnosticTarget()) {
+    for (auto &pendingErrorNote : pendingErrorNotes) {
```

Comment:
This loop is confusing me a bit. Is it fixing all the current diagnostic targets? That seems inefficient. Also, why is that happening when we add a note? Why aren't they correct in the first place? At least, let's break this into it's own function, or move the logic to `reportNoteConsumingDiagnosticError` or something.

Reply:
Agreed on splitting the behavior apart. Sometimes (mostly when importing types) a `SourceLocation` is unavailable, which is why the fixing needs is required. The fixing should happen as close to the usage of the type as possible.